From 64a1abca7b4e8a5fc20db24f036c3df33658a4db Mon Sep 17 00:00:00 2001
From: Beleswar Padhi <b-padhi@ti.com>
Date: Wed, 26 Jun 2024 18:07:57 +0530
Subject: [PATCH 1/6] HACK: remoteproc: k3-r5: Release processor control for
 device operations

Device Operations like get_device, and put_device are processed by
Device Manager. This requires Device Manager to acquire processor
control to execute the operations. Therefore, release processor
control when executing device operations and acquire it back after
the operations are over.

Signed-off-by: Beleswar Padhi <b-padhi@ti.com>
---
 drivers/remoteproc/ti_k3_r5f_rproc.c | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/drivers/remoteproc/ti_k3_r5f_rproc.c b/drivers/remoteproc/ti_k3_r5f_rproc.c
index 57268e7f8ff..81f61cc83c0 100644
--- a/drivers/remoteproc/ti_k3_r5f_rproc.c
+++ b/drivers/remoteproc/ti_k3_r5f_rproc.c
@@ -207,6 +207,8 @@ static int k3_r5f_split_release(struct k3_r5f_core *core)
 
 	dev_dbg(core->dev, "%s\n", __func__);
 
+	k3_r5f_proc_release(core);
+
 	ret = ti_sci_proc_power_domain_on(&core->tsp);
 	if (ret) {
 		dev_err(core->dev, "module-reset deassert failed, ret = %d\n",
@@ -222,6 +224,8 @@ static int k3_r5f_split_release(struct k3_r5f_core *core)
 			dev_warn(core->dev, "module-reset assert back failed\n");
 	}
 
+	ret = k3_r5f_proc_request(core);
+
 	return ret;
 }
 
@@ -445,12 +449,16 @@ static int k3_r5f_split_reset(struct k3_r5f_core *core)
 
 	dev_dbg(core->dev, "%s\n", __func__);
 
+	k3_r5f_proc_release(core);
+
 	if (reset_assert(&core->reset))
 		ret = -EINVAL;
 
 	if (ti_sci_proc_power_domain_off(&core->tsp))
 		ret = -EINVAL;
 
+	ret = k3_r5f_proc_request(core);
+
 	return ret;
 }
 
-- 
2.34.1

